### ISSUE TITLE
py_files option

### ISSUE THREAD

**Issue by davidmarin** [Yelp/mrjob#1375]
Spark's `--py-files` switch is pretty straightforward. Let's support it for streaming steps as well; it's just updating `PYTHONPATH`.

**Comment by davidmarin** [comment]
`--py-file` would make more sense, the way mrjob works (there could be a `py_files` option)

**Comment by davidmarin** [comment]
Oh, looks like `--py-files` might be a bit smarter than that. Still trying to figure out how, when you add a single `.py` file, it makes it available with the correct package and module name.

**Comment by davidmarin** [comment]
So should definitely hold off on making this generally available until I figure out how it works with Spark.

**Comment by davidmarin** [comment]
Okay, it looks like --py-files straight up adds whatever you pass it to `PYTHONPATH`. So passing it a `.py` file does... whatever adding a `.py` file to `PYTHONPATH` does.

Also, unless you're running in cluster mode (`--deploy-mode cluster`), you can only pass local paths to `--py-files`.

**Comment by davidmarin** [comment]
So `py_files` need to do one of three things:

 * in client deploy mode (Hadoop runner), just pass the list of files directly to `--py-files`
 * in cluster deploy mode (EMR runner), upload files to S3 and pass their S3 URIs to `--py-files`
 * in a non-Spark step, upload files to HDFS/S3 (like we would with `upload_files`) and then add them to `PYTHONPATH` in the setup script.

**Comment by davidmarin** [comment]
I don't think it would do any good to allow people to do things like `--py-file my.egg#easter.egg`. It shouldn't make any difference what the file is named anyway.

**Comment by davidmarin** [comment]
Never mind, `--py-files` accepts distributed cache syntax. I think this is something that's hard-wired into Hadoop.

**Comment by davidmarin** [comment]
Here's my script for listing contents of the current working directory:

```python
import os
from mrjob.job import MRJob

class MRSparkFind(MRJob):

    def spark(self, input_path, output_path):
        # Spark may not be available where script is launched
        from pyspark import SparkContext

        sc = SparkContext(appName='mrjob Spark find')

        paths = [p for _, __, paths in os.walk('.') for p in paths]

        rdd = sc.parallelize(paths)
        rdd.saveAsTextFile(output_path)

        sc.stop()


if __name__ == '__main__':
    MRSparkFind.run()
```

**Comment by davidmarin** [comment]
Oh, this is dumb:

Looks like `--py-files` accepts distributed cache syntax, but then pastes the whole thing, hash and all, into `PYTHONPATH`. You *can* use distributed cache syntax to control the name of the Python file, but then it won't appear in your `PYTHONPATH` in a usable way.

This is true in Spark 1.3.1 at least (the version on AMI 3.11.0). This is a rather old version of Spark, and this bug might have been fixed since then.

**Comment by davidmarin** [comment]
Fixed by #1466.
